full-service/src/service/gift_code.rs

```diff
@@ -0,0 +1,644 @@
+// Copyright (c) 2020 MobileCoin Inc.
+
+//! Service for managing gift codes.
+//!
+//! Gift codes are onetime accounts that contain a single TXO. They provide
+//! a means to send MOB in a way that can be "claimed," for example, by pasting
+//! a QR code for a gift code into a group chat, and the first person to
+//! consume the gift code claims the MOB.
+
+use crate::{
+    db::{
+        account::{AccountID, AccountModel},
+        b58_encode,
+        gift_code::{GiftCodeDbError, GiftCodeModel},
+        models::{Account, GiftCode, TransactionLog, Txo, TX_FAILED, TX_PENDING, TX_SUCCEEDED},
+        transaction_log::TransactionLogModel,
+        txo::TxoModel,
+    },
+    error::{WalletDbError, WalletServiceError},
+    service::{
+        decorated_types::{JsonGiftCode, JsonSubmitResponse},
+        WalletService,
+    },
+};
+use mc_account_keys::{AccountKey, RootIdentity};
+use mc_common::logger::log;
+use mc_connection::{BlockchainConnection, UserTxConnection};
+use mc_crypto_keys::{CompressedRistrettoPublic, RistrettoPublic};
+use mc_fog_report_connection::FogPubkeyResolver;
+use mc_ledger_db::Ledger;
+use mc_transaction_core::{constants::MINIMUM_FEE, get_tx_out_shared_secret};
+use mc_util_from_random::FromRandom;
+
+use diesel::prelude::*;
+use displaydoc::Display;
+use std::convert::TryFrom;
+
+#[derive(Display, Debug)]
+pub enum GiftCodeServiceError {
+    /// Building the gift code failed
+    BuildGiftCodeFailed,
+
+    /// Unexpected TxStatus while polling: {0}
+    UnexpectedTxStatus(String),
+
+    /// Gift Code transaction produced an unexpected number of outputs: {0}
+    UnexpectedNumOutputs(usize),
+
+    /// Gift Code does not contain enough value to cover the fee: {0}
+    InsufficientValueForFee(u64),
+
+    /// Unexpected number of Txos in the Gift Code Account: {0}
+    UnexpectedNumTxosInGiftCodeAccount(usize),
+
+    /// Unexpected Value in Gift Code Txo: {0}
+    UnexpectedValueInGiftCodeTxo(u64),
+}
+
+pub trait GiftCodeService {
+    /// Builds a new gift code.
+    ///
+    /// Building a gift code requires the following steps:
+    ///  1. Create a new account to receive the funds
+    ///  2. Send a transaction to the new account
+    ///  3. Wait for the transaction to land
```

Comment:
It doesn't look like the implementation of `build_and_submit_gift_code` is waiting for anything, it just submits the transaction.